### PR TITLE
Add optional notifications support through SNS

### PR DESCRIPTION
**Pull request by diego-ojeda-binbash**
Optional notifications support was added. It can create an SNS topic and add subscriptors to it.
This functionality is useful to be notified about successful/failed backups, restore jobs, copy jobs and more backup-related events.

### CODE REVIEW

**Review by exequielrafaela — APPROVED** [review]

**Comment by exequielrafaela** [comment]
@diego-ojeda-binbash OJ If I've understood correctly, on the one hand, this new feature creates the SNS topic with all the necessary resources so a subscriber can get the notification. On the other hand, it would be very useful to have the possibility to **optionally** pass a pre-existing SNS topic arn like the one resulting from https://github.com/binbashar/terraform-aws-notify-slack that we frequently use. Equivalent to our CloudTrail and Config notifications setup, which would probably be the approach we will take for the backup service too.

I've already approved it ✅, may be we could extend it in this PR, or a new one.

Thanks for your continuous and effortless collaboration!!! 🚀 💪🏼 👏🏼

**Comment by lgallard** [comment]
@diego-ojeda-binbash I'd rather separate resources not related to the main function (in this cases AWS Backup) and instead pass the "external" resources (in this example, the SNS topic arn) as an module input. 

Why? It's more difficult to keep track of resources changes/evolution of each resource and fix or enhance them.

To be honest, I like the idea of adding a SNS topic in case you don't have one, but I agree with @exequielrafaela on the fact that it would be useful to pass an external SNS topic arn in case you had already created it from other services / logic.

File: notifications.tf [hunk 1]
```diff
@@ -0,0 +1,48 @@
+locals {
+    # Whether to enable notifications or not
+    enable_notifications = var.enabled && length(var.notifications_backup_vault_events) > 0 ? true : false
+}
+
+resource "aws_sns_topic" "backup_events" {
```
Comment (lgallard) [inline notifications.tf:6]: I'd rather leave all the SNS code outside the module

File: notifications.tf [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+resource "aws_sns_topic_policy" "backup_events" {
+  count  = local.enable_notifications ? 1 : 0
+  arn    = aws_sns_topic.backup_events[0].arn
+  policy = data.aws_iam_policy_document.backup_events.json
+}
+
+resource "aws_backup_vault_notifications" "backup_events" {
+  count               = local.enable_notifications ? 1 : 0
+  backup_vault_name   = var.vault_name != null ? var.vault_name : "Default"
+  sns_topic_arn       = aws_sns_topic.backup_events[0].arn
```
Comment (lgallard) [inline notifications.tf:37]: Once the SNS topic was created outside this module, pass the ARN here, using the SNS resource or module output as the input of this module

File: variables.tf [hunk 3]
```diff
@@ -142,3 +142,31 @@ variable "windows_vss_backup" {
   type        = bool
   default     = false
 }
+
+#
+# Notifications
+#
+variable "notifications_backup_vault_events" {
```
Comment (lgallard) [inline variables.tf:149]: I agree on this variable definition

File: variables.tf [hunk 4]
```diff
@@ -142,3 +142,31 @@ variable "windows_vss_backup" {
...
+  description = "A list of backup events that you would like to be notified about. Leave it empty to disable notifications."
+  type        = list
+  default     = []
+}
+
+variable "notifications_sns_topic_name" {
```
Comment (lgallard) [inline variables.tf:155]: But the SNS variable would go outside the module

**Review by lgallard — CHANGES_REQUESTED** [review]

**Comment by diego-ojeda-binbash** [comment]
@exequielrafaela @lgallard Thanks for all your comments! After careful thinking I believe it would be best to move this functionality to a separate module where can also handle the use case that @exequielrafaela suggested. That way, the new module will be able to handle both fan-in and fan-out messaging strategies instead of being constrained to only one of them.

**Comment by lgallard** [comment]
@diego-ojeda-binbash @exequielrafaela release 0.7.0 has an implementation based on this PR. Now you can define notifications as follows:

```
module "aws_backup_example" {

  source = "lgallard/backup/aws"

  # Vault
  vault_name = "vault-3"

  # Plan
  plan_name = "complete-plan"

  # Notifications
  notifications = {
    sns_topic_arn       = aws_sns_topic.backup_vault_notifications.arn
    backup_vault_events = ["BACKUP_JOB_STARTED", "BACKUP_JOB_COMPLETED", "BACKUP_JOB_FAILED", "RESTORE_JOB_COMPLETED"]
  }
```

Use `sns_topic_arn` to pass an already created SNS topic by a resource or a module as mentioned by @exequielrafaela (/terraform-aws-notify-slack module)

**Pull request closed without merging**